monkey/agent_plugins/credentials_collectors/chrome/src/windows_credentials_database_processor.py

```diff
@@ -1,15 +1,133 @@
-from pathlib import PurePath
-from typing import Sequence
+import logging
+import os
+import shutil
+import sqlite3
+import string
+import tempfile
+from pathlib import PureWindowsPath
+from typing import List, Optional, Sequence, Set, Tuple
 
-from common.credentials import Credentials
+from Crypto.Cipher import AES
+
+from common.credentials import Credentials, Password, Username
 from common.types import Event
+from common.utils.code_utils import secure_generate_random_string
 
+from .windows_utils import Win32CryptUnprotectData
+
+logger = logging.getLogger(__name__)
 
-class WindowsCredentialsDatabaseProcessor:
-    def __init__(self):
-        pass
 
+# TODO: use pathlib.Path, not str
+class WindowsCredentialsDatabaseProcessor:
     def __call__(
-        self, interrupt: Event, database_paths: Sequence[PurePath]
+        self, interrupt: Event, database_paths: Set[Tuple[PureWindowsPath, Optional[bytes]]]
     ) -> Sequence[Credentials]:
-        return []
+        credentials = []
+
+        # TODO: make interruptible
+        for database_path, master_key in database_paths:
+            # copy database before querying it to bypass lock errors
+            path = self._copy_db(database_path)
+            if path:
+                try:
+                    credentials.extend(self._export_credentials(path, master_key))
+                except Exception as err:
+                    logger.debug(f"{err}")
+
+                self._clean_file(path)
+
+        return [
+            # TODO: fix assumption that identity is a username, could be an email ID;
+            #       or just let it be, anything that uses a `Username` which is actually
+            #       an email ID will just fail and move on
+            Credentials(identity=Username(username=username), secret=Password(password=password))
+            for username, password in set(credentials)
+        ]
+
+    def _copy_db(self, database_path) -> Optional[str]:
+        root_dir = [
+            tempfile.gettempdir(),
+            # NOTE: (This is not relevant ATM since we only steal credentials from
+            # the current user, but leave this comment here for future reference.)
+            # Using user tempfile will produce an error when impersonating users
+            # (permission denied). Use a public directory in that case (uncomment the following):
+            # os.environ.get("PUBLIC", None),
+            # os.environ.get("SystemDrive", None) + "\\",
+        ]
+        random_name = secure_generate_random_string(n=9, character_set=string.ascii_lowercase)
+
+        for r in root_dir:
+            try:
+                temp = os.path.join(r, random_name)
```

Comment:
Just use `tempfile.mkdtemp()`

Reply:
But, consider using `tempfile.TemporaryFile()`. It will automatically be removed when the object is garbage collected.

Reply:
That'd require us to modify the database copying logic when we implement stealing credentials from other users. I'm not sure how bad it is that we're handling the removal of the file ourselves.